waku/v2/protocol/waku_store/waku_store_types.nim

```diff
@@ -368,20 +370,28 @@ proc new*(T: type StoreQueueRef, capacity: int): T =
   return StoreQueueRef(items: items, capacity: capacity)
 
 proc add*(storeQueue: StoreQueueRef, msg: IndexedWakuMessage): StoreQueueResult[void] =
-  ## Add a message to the queue.
+  ## Add a message to the queue
   ## If we're at capacity, we will be removing,
   ## the oldest (first) item
   
+  # Ensure that messages don't "jump" to the front of the queue with future timestamps
+  if msg.index.senderTime - msg.index.receiverTime > MaxTimeVariance:
+    return err("future_sender_timestamp")
```

Comment:
Thanks! It's good to have defensive sanity checks, but I think here we'd have had to introduce multiple bugs for this to happen, because of [how we generate receiver timestamps](https://github.com/status-im/nim-waku/blob/d2fccb5220144893f994a67f2cc26661247f101f/waku/v2/protocol/waku_store/waku_store.nim#L60) which should always result in a reasonable and positive value.

Reply:
AFAIU the buffer is not authenticated, so an attacker may actually craft a payload that crashes the implementation at that point. I might be wrong though.

Reply:
I meant the corresponding protobuffer field. Can an attacker specifically craft one with a negative timestamp? Since PB are not authenticated, it seems so to me. With a quick look, it seems that here https://github.com/status-im/nim-waku/blob/4e5dfe66400e46cf0a4f5ede12a5a6a45b109b62/waku/v2/protocol/waku_store/waku_store.nim#L368 the index is computed from the input parameter receiverTime, which is read from a WakuMessageStore database (int64) in https://github.com/status-im/nim-waku/blob/4e5dfe66400e46cf0a4f5ede12a5a6a45b109b62/waku/v2/node/storage/message/waku_message_store.nim#L187 (few lines earlier there is how a record enters such database) However, such Index cursor timestamp seems to be initialized directly from the (non-authenticated) protobuffer encoding the message: https://github.com/status-im/nim-waku/blob/4e5dfe66400e46cf0a4f5ede12a5a6a45b109b62/waku/v2/protocol/waku_store/waku_store.nim#L124

Reply:
> Can an attacker specifically craft one with a negative timestamp?

Not as far as I can see, unless the attacker gains access to your local SQLite database and injects data there.
Use cases:
1. `relay` messages being stored the first time: an index is computed for these messages and the message is added according to the `add` logic in this PR:
https://github.com/status-im/nim-waku/blob/4e5dfe66400e46cf0a4f5ede12a5a6a45b109b62/waku/v2/protocol/waku_store/waku_store.nim#L404-L405
When the index gets computed a new receiverTime is generated:
https://github.com/status-im/nim-waku/blob/4e5dfe66400e46cf0a4f5ede12a5a6a45b109b62/waku/v2/protocol/waku_store/waku_store.nim#L60

2. When `store` node gets initialised, it reads messages from the SQLite DB and adds it using the same `add` proc as above, except this time the `receiverTime` is not generated again, but read from what was stored before. Note that the origin of all these timestamps would still be the generation from (1), unless someone modified the SQLite DB (which would cause a different type of security concern). This is what you linked in your first two snippets above.

3. `receiverTime` is also used in Indices used as cursors in HistoryRequests and HistoryResponses - this is what gets serialised and deserialised in the protobuf objects you linked. These can indeed be modified and aren't authenticated, but is never added to the internal store: we just use them for lookups. If someone were to inject invalid timestamps into the `cursor`, their history query would simply fail.

If we're still not on the same page, let's set up a call. Maybe some clarification will help. :)

Reply:
I miss the big picture and how parts combine together, so take the followings with care. 

Regarding 3. If I change https://github.com/status-im/nim-waku/blob/a8cf71e361990fd9c5d3e659b6626c9101acd732/tests/v2/test_waku_store.nim#L579 to

`msgList = @[WakuMessage(payload: @[byte 0], contentTopic: ContentTopic("2"), timestamp: Timestamp(int64.low)),`

(which I guess tests the kind of queries you're referring to), I receive "Unhandled exception: over- or underflow [OverflowError]". 

If I do a query using int64.low and int64.high as time window, i.e. I change https://github.com/status-im/nim-waku/blob/a8cf71e361990fd9c5d3e659b6626c9101acd732/tests/v2/test_waku_store.nim#L682

to

`      let rpc = HistoryQuery(contentFilters: @[HistoryContentFilter(contentTopic: ContentTopic("1"))], startTime: Timestamp(int64.low), endTime: Timestamp(int64.high))`

again I receive `Unhandled exception: over- or underflow [OverflowError]`, and so on.

I observe similar behaviors if you specify `timestamp: int64.low ` when a WakuMessage is initialized (but since in other tests this field is not explicitly passed, I didn't investigate further since I don't know if it the timestamp is read from a received message or is set at the application level).

Now, these examples might still be artificial and in practice are never triggered. However such sanity checks are almost for free (i.e. a couple of extra ifs and no huge computation is involved), so IMO it is better to have them rather than not, especially if inputs can come from untrusted sources (and considering also that new code may be introduced somewhere else in the future that can make such unchecked conditions _bugs_).